firewood/src/merkle.rs

```diff
@@ -1564,14 +2089,14 @@ mod tests {
 
     #[test_case(Bincode::new(), leaf(Vec::new(), Vec::new()) ; "empty leaf encoding with Bincode")]
     #[test_case(Bincode::new(), leaf(vec![1, 2, 3], vec![4, 5]) ; "leaf encoding with Bincode")]
-    #[test_case(Bincode::new(), branch(Some(b"value".to_vec()), vec![1, 2, 3].into()) ; "branch with chd with Bincode")]
-    #[test_case(Bincode::new(), branch(Some(b"value".to_vec()), None); "branch without chd with Bincode")]
-    #[test_case(Bincode::new(), branch(None, None); "branch without value and chd with Bincode")]
+    #[test_case(Bincode::new(), branch(b"", b"value", vec![1, 2, 3].into()) ; "branch with chd with Bincode")]
+    #[test_case(Bincode::new(), branch(b"", b"value", None); "branch without chd with Bincode")]
+    #[test_case(Bincode::new(), branch_without_data(b"", None); "branch without value and chd with Bincode")]
     #[test_case(PlainCodec::new(), leaf(Vec::new(), Vec::new()) ; "empty leaf encoding with PlainCodec")]
     #[test_case(PlainCodec::new(), leaf(vec![1, 2, 3], vec![4, 5]) ; "leaf encoding with PlainCodec")]
-    #[test_case(PlainCodec::new(), branch(Some(b"value".to_vec()), vec![1, 2, 3].into()) ; "branch with chd with PlainCodec")]
-    #[test_case(PlainCodec::new(), branch(Some(b"value".to_vec()), Some(Vec::new())); "branch with empty chd with PlainCodec")]
-    #[test_case(PlainCodec::new(), branch(Some(Vec::new()), vec![1, 2, 3].into()); "branch with empty value with PlainCodec")]
+    #[test_case(PlainCodec::new(), branch(b"", b"value", vec![1, 2, 3].into()) ; "branch with chd with PlainCodec")]
+    #[test_case(PlainCodec::new(), branch(b"", b"value", Some(Vec::new())); "branch with empty chd with PlainCodec")]
+    #[test_case(PlainCodec::new(), branch(b"", b"", vec![1, 2, 3].into()); "branch with empty value with PlainCodec")]
```

Comment:
Rather than rolling each permutation by hand, the `test_case` crate also has `test_matrix` which can test all the combinations for you:
```
   #[test_matrix(
        [b"", b"key"],
        [b"", b"value"],
        [None, Some(vec![1,2,3])]
    )]
    fn encodes_then_decode_branches(key: &'static [u8], value: &'static [u8], vec: Option<Vec<u8>>) {
        encode(branch(key, value, vec));
    }

    #[test_matrix(
        [vec![], vec![1, 2. 3]],
        [vec![], vec![1, 2, 3]]
    )]
    fn encode_then_decode_leaves(key: Vec<u8>, value: Vec<u8>) {
        encode(leaf(key, value))
    }
```

Reply:
See issue #512 for the `test-matrix` comment. 

@xinifinity 
>It doesn't seem we have a test case for branch with non empty path?

Correct, the new encoding/decoding implementation isn't currently used anywhere. Can we push the implementation out after this PR lands? It's impossible that this doesn't get caught because all the proofs will start failing once we switch to the new encoding/decoding. It should also be trivial to implement